design/history/exploration-reports/2020.10-data-model-numbers-and-codecs.md

```diff
@@ -0,0 +1,77 @@
+IPLD Data Model numbers and Ipld Codecs
+=======================================
+
+**Author**: Volker Mische ([@vmx])
+
+This is an exploration report about how to serialize [IPLD Data Model] numbers into bytes and back.
+
+
+Canonical representations
+-------------------------
+
+In IPLD it's desirable to have one single representation for every instance of the Data Model. This would lead to a bijective mapping between the Data Model and the encoded bytes. If you have a Data Model instance and you use a specific codec, it should always result in the same bytes, independent of the programming language used. That's important for content-addressing.
+
+
+### Conversion into the Data Model
+
+This looks like a hard problem, e.g. if you look at this IPFS issue with [DAG-CBOR] where there is a [different encoding between JavaScript and Go when numbers are integers]. Though in this issue, the problem is **not** the Data Model to Codec conversion. The problem is the **input data to Data Model conversion**. It took me a long time to figure that out.
+
+When I think of the IPLD Data Model I often think in terms of JSON and having numbers represented as text. I then think about how to represent those things as bytes. If you think about it that way, you have problems like: is `4251.00` and `4251` the same number? If yes, then it should be encoded the same way?
+
+From a Data Model perspective that's a non-issue. You have two kinds of numbers: [Integer] and [Float]. You can clearly distinguish between `Integer(4251)` and `Float(4251)`.
```

Comment:
Should we consider changing this? We could just say “numbers” instead of integer and float. At the end of the day, the data model just tries to de-serialize into common types, but if a specific type difference is required in a specific language and application that’s going to be mapped on above our Data Model. It’s not in conflict with the Data Model to de-serialize all integers as BigInt if that’s what the application requires, nor is in conflict to convert a list of integers to floats if what you need to do on them is floating point math. It may be better/simpler to say “number” is the Data Model type and leave it up to languages to figure out how each serialization maps to their type system.

That would leave us free to say “there is only one way to encode 4251 and you can’t do it as a float” in formats we design for a very high degree of hash consistency. **Formats** may still distinguish between integers and floats, and that’s fine, but the Data Model doesn’t have to.

Reply:
That's a strong "no we don't change this" from me. It is soooo good that we have this distinction in the Data Model. JavaScript is the one who is wrong here. I'd hope that this exploration report makes that clear. In case it doesn't I'm happy to see if I can put together a follow up report.

Reply:
I’m not going to try and defend JavaScript’s number type system, it’s fairly indefensible, but for me this isn’t about JavaScript and more about what the data model’s view of “lossless” means.

Finding the smallest binary representation of a number is a lossy operation from the point of view of almost any number type system. Even VARINT loses some type information you may have had on your numbers that you’ll have to add back if you want to maintain the same integer type. For instance, if you have a 64b integer type of a small number value it’s probably going to round-trip to a smaller number type that you’ll then need to convert back to 64b.

Depending on the language, it’s often going to be necessary to do some form of type conversion to satisfy the needs of the application that is working with the de-serialized state.

So the question is, does unifying these into a single number type incur a form of loss that makes the type effectively unusable? That answer might be yes, I’m leaning towards yes (in agreement with you), but I want to explore exactly why.

At the end of that day, this is all just serialized binary data that we’re applying logic to in order to convert to useful types. With integers, we have pretty simple and recurrent patterns for working with any loss the format may cause during serialization. I don’t think we have that for floats, and as I keep being reminded, floats are **NOT INTEGERS** and can’t be represented effectively if you’re trying to normalize and reconcile them with integers. They are a very different type from integers in **every** programming language, including JavaScript ([Float32Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Float32Array)) , because there’s just no way for a language to work around the fact that the math operations for them are implemented by a separate set of instructions.

Reply:
> Finding the smallest binary representation of a number is a lossy operation

Yes. It depends on the definition of "lossy". I wouldn't call it lossy, I would call it mutation or transformation. Though I totally get what you mean.

>  if you want to maintain the same integer type

This exploration report is about the Data Model only. And that's important here. I've also mixed the Data Model end native language encoding in my mind in the past, but I think we shouldn't. If you think about the Data Model only, there are no different integer types, there is only a single one.

> floats are **NOT INTEGERS** and can’t be represented effectively if you’re trying to normalize and reconcile them with integers

I'm not sure what you mean. Converting an integer to a varint is the same kind of transformation as if you convert it to a double (keeping the bit size limitations in mind, e.g. < 53 bit for a 64-bit float).

Reply:
This is really about where we want to draw a line between utility and fidelity. We can use the same logic in this discussion to further explore:

 * Why do we bother with a `String` type? They're all just byte arrays in the end and strings are only really interesting in that they can represent things that humans can read. They're essentially sugar with weird rules for conversion (utf-8).
 * Why bother with a `Boolean` type? C has made do with `0` and `1` forever, JavaScript makes good use of "truthy" and "falsy" in the same spirit.
 * Maps really are a pain, we should rule them out in the representation layer and rely on Schemas to _transform_ them for users from tuple representation (and other) into maps in memory. This is what Filecoin is doing after all and it really does solve a lot of problems and saves space when encoded.

I think it's legitimate to argue for `Number` over `Int` and `Float` as long as we're boiling down all of `int8`, `int16`, `int32`, `int64`, `uint8`, `uint16`, `uint32`, `uint64` into a single type, why not throw `float16`, `float32` and `float64` into the same basket? I know @warpfork's argument here is about cardinality - when you treat `1.0` and `1` as the same thing you reduce cardinality and lose information. But you're doing that anyway when you reduce all of the int types to `Int`.

But from a utility perspective, I think we JavaScript programmers should be the ones arguing that greater utility lies in differentiating between integers and floats, `x % 1 === 0` _is a **really** crappy thing to have to do_ (same goes for `Number.isInteger()` which is just sugar over 💩) and `Number.MAX_SAFE_INTEGER` is equally crappy.

"My system deals with floating point numbers in _this_ place and deals with integers in _that_ place" is pretty useful. I also like that we can shove floats into their own box and push it off to the side, and recommend that users to likewise with serialisation. `Number` is just a lowest common denominator which ends spoiling integers.